File: tests/fixtures/scenario_generator/network_protocol_impl/decoder.c
/* fixture for network_protocol_impl — C-language stub.
 *
 * Present so the network_protocol_impl fingerprint, which requires a
 * .c AND a .rs AND a .go file to all match magic strings, can fire.
 * The compiled code is irrelevant — only the strings parse_packet,
 * encode_packet, and packet_header in the source are load-bearing
 * for detection.
 *
 * The discoverer also finds protocol_encode/protocol_decode and any
 * dispatch table here.
 */

/* packet_header — wire-format prelude struct. */
struct packet_header {
    unsigned int magic;
    unsigned int length;
};

/* protocol_decode — discoverer entry point (matches *_decode). */
int protocol_decode(const unsigned char *buf, unsigned int len)
{
    (void)buf;
    return (int)len;
}

/* protocol_encode — discoverer entry point (matches *_encode). */
int protocol_encode(unsigned char *out, const struct packet_header *h)
{
    out[0] = (unsigned char)h->magic;
    return 1;
}

/* parse_packet — also matches the network_protocol_impl fingerprint string. */
int parse_packet(const unsigned char *buf, unsigned int len)
{
    return protocol_decode(buf, len);
}

/* encode_packet — also matches the network_protocol_impl fingerprint string. */
int encode_packet(unsigned char *out, const struct packet_header *h)
{
    return protocol_encode(out, h);
}

File: tests/fixtures/scenario_generator/network_protocol_impl/decoder.rs
// fixture for network_protocol_impl — Rust-language stub.
//
// Present so the network_protocol_impl fingerprint (which requires
// the strings parse_packet AND PacketHeader to appear in a *.rs file)
// can fire. The encoder/decoder/handler functions below are also
// picked up by the network_protocol_impl discoverer.
//
// Lives standalone (no Cargo.toml in this fixture) so neither
// library_rust ([lib]) nor cli_rust ([[bin]]) can also match.

/// PacketHeader — wire-format prelude.
pub struct PacketHeader {
    pub magic: u32,
    pub version: u32,
    pub length: u32,
}

/// parse_packet — also matches the network_protocol_impl fingerprint string.
pub fn parse_packet(data: &[u8]) -> Option<PacketHeader> {
    if data.len() < 12 {
        return None;
    }
    Some(PacketHeader {
        magic: u32::from_be_bytes([data[0], data[1], data[2], data[3]]),
        version: u32::from_be_bytes([data[4], data[5], data[6], data[7]]),
        length: u32::from_be_bytes([data[8], data[9], data[10], data[11]]),
    })
}

/// decode_header — discoverer entry point (matches decode_*).
pub fn decode_header(data: &[u8]) -> Option<PacketHeader> {
    parse_packet(data)
}

/// encode_header — discoverer entry point (matches encode_*).
pub fn encode_header(h: &PacketHeader, out: &mut [u8]) {
    out[0..4].copy_from_slice(&h.magic.to_be_bytes());
    out[4..8].copy_from_slice(&h.version.to_be_bytes());
    out[8..12].copy_from_slice(&h.length.to_be_bytes());
}

/// handle_data_packet — discoverer entry point (matches handle_*_packet).
pub fn handle_data_packet(h: &PacketHeader, payload: &[u8]) -> bool {
    let _ = h;
    let _ = payload;
    true
}
